onnxruntime/core/providers/cpu/tensor/scatter.cc

```diff
@@ -0,0 +1,211 @@
+        .TypeConstraint("T", DataTypeImpl::AllTensorTypes())
+        .TypeConstraint("Tind", std::vector<MLDataType>{DataTypeImpl::GetTensorType<int32_t>(), DataTypeImpl::GetTensorType<int64_t>()}),
+    Scatter);
+
+template <class Tin>
+Status CopyScatterData(const Tensor* data_input, const Tensor* indices_input, const Tensor* updates_input,
+                       const int64_t axis, Tensor* data_output) {
+  const TensorShape& input_data_shape = data_input->Shape();
+  const Tin* indices_data = indices_input->template Data<Tin>();
+  const auto num_indices = indices_input->Shape().Size();
+  for (int64_t i = 0; i < num_indices; ++i) {
+    Tin idx = indices_data[i];
+    if (idx < 0 || idx >= input_data_shape[axis]) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "indices element out of data bounds, idx=", idx,
+                             " data_dim=", input_data_shape[axis]);
+    }
+  }
+
+  const auto input_elements = input_data_shape.Size();
+  const auto element_bytes = data_input->DataType()->Size();
+  const auto total_input_bytes = data_input->Size();
+
+  const uint8_t* src_base = reinterpret_cast<const uint8_t*>(data_input->DataRaw());
+  uint8_t* dst_base = reinterpret_cast<uint8_t*>(data_output->MutableDataRaw());
+  const bool is_string_type = data_input->DataType() == DataTypeImpl::GetType<std::string>();
+
+  // We allow runtime to re-use input for output. If input/output Tensor* are the same
+  // we do not copy
+  if (src_base != dst_base) {
+    if (is_string_type) {
+      const std::string* str_begin = data_input->template Data<std::string>();
+      const std::string* str_end = str_begin + input_elements;
+      std::string* dst = data_output->template MutableData<std::string>();
+      std::copy(str_begin, str_end, dst);
+    } else {
+      memcpy(dst_base, src_base, total_input_bytes);
+    }
+  }
+
+  // Now poke updates
+
+  const auto& upd_shape = updates_input->Shape();
+  const auto num_dims = input_data_shape.NumDimensions();
+  assert(num_dims > 0);
+
+  // Allocate and zero out counts. The input/output is of the same rank as
+  // indices/updates but the actual dimensions of indices/updates must be less or equal
+  // than that of input/output because we can update no more elements than
+  // the input contains. As we walk through the indices/updates
+  // we maintain dimension count as we will need to use it
+  // to compute output offset but using input/output dim values.
+  // We treat the whole array as a number where each element having
+  // different cardinality according to the upd_shape dimensions.
+  // As each counter reaches its max (upd_shape) it resets to zero
+  // and we carry to the more significant dim (right to left)
+  std::vector<int64_t> dim_counters(num_dims);
+
+  // This vector contains number of lower dimension elements per
+  // unit. For example, for the dimensions of [2, 3] the vector
+  // would contain [3, 1] since for each count of dim 0 it
```

Comment:
Would the example be clearer if there were 3 dims?

If I understand correctly, for dimensions of [2, 3, 4] the vector would contain [7, 4, 1]. Is that correct?